src/3d/qgs3dexportobject.cpp

```diff
@@ -0,0 +1,159 @@
+{
+  // TODO: delete vertices that are not used
+  for ( int i = 0; i < positionsBuffer.size(); i += 3 )
+  {
+    for ( int j = 0; j < 3; ++j )
+    {
+      mVertxPosition << positionsBuffer[i + j] * scale + translation[j];
+    }
+  }
+
+  for ( int i = 0; i < faceIndex.size(); i += 3 )
+  {
+    // skip invalid triangles
+    if ( faceIndex[i] == faceIndex[i + 1] && faceIndex[i + 1] == faceIndex[i + 2] ) continue;
+    for ( int j = 0; j < 3; ++j ) mIndexes << faceIndex[i + j] + 1;
+  }
+}
+
+void Qgs3DExportObject::setupNormalCoordinates( const QVector<float> &normalsBuffer )
+{
+  mNormals << normalsBuffer;
+}
+
+void Qgs3DExportObject::setupTextureCoordinates( const QVector<float> &texturesBuffer )
+{
+  mTexturesUV << texturesBuffer;
+}
+
+void Qgs3DExportObject::objectBounds( float &minX, float &minY, float &minZ, float &maxX, float &maxY, float &maxZ )
+{
+  for ( int vertice : mIndexes )
+  {
+    int heightIndex = ( vertice - 1 ) * 3 + 1;
+    minX = std::min( minX, mVertxPosition[heightIndex - 1] );
+    maxX = std::max( maxX, mVertxPosition[heightIndex - 1] );
+    minY = std::min( minY, mVertxPosition[heightIndex] );
+    maxY = std::max( maxY, mVertxPosition[heightIndex] );
+    minZ = std::min( minZ, mVertxPosition[heightIndex + 1] );
+    maxZ = std::max( maxZ, mVertxPosition[heightIndex + 1] );
+  }
+}
+
+void Qgs3DExportObject::saveTo( QTextStream &out, float scale, const QVector3D &center )
+{
+  // Set groups
+  // turns out grouping doest work as expected in blender
+
+  // smoothen edges
+  if ( mSmoothEdges ) out << "s on\n";
+  else out << "s off\n";
+
+  // Construct vertices
+  for ( int i = 0; i < mVertxPosition.size(); i += 3 )
+  {
+    // for now just ignore wrong vertex positions
+    out << "v ";
+    out << ( mVertxPosition[i] - center.x() ) / scale << " ";
+    out << ( mVertxPosition[i + 1] - center.y() ) / scale << " ";
+    out << ( mVertxPosition[i + 2] - center.z() ) / scale << "\n";
+    if ( i + 3 <= mNormals.size() )
+    {
+      out << "vn " << mNormals[i] << " " << mNormals[i + 1] << " " << mNormals[i + 2] << "\n";
+    }
+    int u_index = i / 3 * 2;
+    if ( u_index + 1 < mTexturesUV.size() )
+    {
+      // TODO: flip texture in a more appropriate way (for repeated textures)
+      out << "vt " << mTexturesUV[u_index] << " " << 1.0f - mTexturesUV[u_index + 1] << "\n";
+    }
+  }
+
+  bool hasTextures = mTexturesUV.size() == mVertxPosition.size() / 3 * 2;
+  // if the object has normals then the normals and positions buffers should be the same size
+  bool hasNormals = mNormals.size() == mVertxPosition.size();
+
+  if ( !hasNormals && !mNormals.empty() )
+  {
+    qDebug() << "WARNING: vertex normals count and vertex positions count are different";
+  }
+  int verticesCount = mVertxPosition.size() / 3;
+
+  auto getVertexIndex = [&]( int i ) -> QString
+  {
+    int negativeIndex = -1 - ( verticesCount - i );
```

Comment:
why are the indices negative here? something that .obj format requires?

Reply:
In the .obj spec negative indices refer to the last defined vertices (-1 for the last vertice).
This helps keeping each object data separated